### PR TITLE
Make plot_gate_map generic for any backend

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Previously the plot_gate_map() family of visualization functions only
worked with a fixed hard coded list of number of qubits. The function
made the assumption if the backend had a matching number of qubits it
was an IBM backend that had the same number of qubits (and fell into a
fixed connectivity topology) and would use the manually set graph layout
to visualize the function. This was problematic because it didn't
actually work for any backend, and if a backend had the same number of
qubits as an IBM backend but a different connectivity it would use a potentially
nonoptimal layout. This also required us to update the hard
coded list whenever a new IBM backend was released to support
visualizing them. Additionally the functions only worked with BackendV1
and BaseBackend based backends and would fail if a BackendV2 object were
passed to the function. This commit address these issues by making 4 key
changes to the function, first it adds a new qubit_coordinates kwarg to
all the methods which if specifed can be used to manually specify a
graph layout to use for the visualization. The second is that the
hardcoded list for layouts are only used if it's an IBM backend (or a
fake backend) so that we're using them where appropriate (the 127q
layout for ibm_washington is also added). The third change is that if a
graph layout isn't provided and it's not an ibm backend with a hardcoded
entry, retworkx's spring_layout() function is used to create a layout.
In the future we can use a more appropriate planar_layout() function
after its available in retworkx (see
https://github.com/Qiskit/retworkx/issues/438 ). Finally the last change
is that BackendV2 support is added to the function to make sure we're
able to use the visualization function regardless of what type of
backend is passed in.

### Details and comments

### CODE REVIEW

File: qiskit/visualization/gate_map.py [hunk 1]
```diff
@@ -338,14 +353,182 @@ def plot_gate_map(
...
+        scaling_factor = 10 ** int(math.log10(num_qubits) + 1)
+        qubit_coordinates = [
+            (
+                int(scaling_factor * qubit_coordinates_rx[i][0]),
+                int(scaling_factor * qubit_coordinates_rx[i][1]),
+            )
+            for i in range(num_qubits)
+        ]
+
+    if any(x[0] < 1 or x[1] < 1 for x in qubit_coordinates):
```
Comment (kevinhartman) [inline qiskit/visualization/gate_map.py:515]: Should these check if < 0 instead? If the smallest value is 0, won't `qubit_coordinates` be unchanged?

File: qiskit/visualization/gate_map.py [hunk 2]
```diff
@@ -705,24 +908,78 @@ def plot_error_map(backend, figsize=(12, 9), show_title=True):
...
+                        directed = True
+                        break
+
+            for line in cmap:
+                for item in props["gates"]:
+                    if item["qubits"] == line:
+                        cx_errors.append(item["parameters"][0]["value"])
+                        break
+                else:
+                    continue
```
Comment (kevinhartman) [inline qiskit/visualization/gate_map.py:945]: I see this was here originally, but is there any point to this `continue`? As I understand it, it should advance the outer loop (`for line in map`), but there's nothing left to do anyway.

File: qiskit/visualization/gate_map.py [hunk 3]
```diff
@@ -705,24 +908,78 @@ def plot_error_map(backend, figsize=(12, 9), show_title=True):
...
+        for gate, prop_dict in backend.target.items():
+            if prop_dict is None or None in prop_dict:
+                continue
+            for qargs, inst_props in prop_dict.items():
+                if gate == "measure":
+                    if inst_props.error is not None:
+                        read_err[qargs[0]] = inst_props.error
+                elif len(qargs) == 1:
+                    if inst_props.error is not None:
+                        single_gate_errors = max(single_gate_errors[qargs[0]], inst_props.error)
```
Comment (kevinhartman) [inline qiskit/visualization/gate_map.py:967]:
```suggestion
                        single_gate_errors[qargs[0]] = max(single_gate_errors[qargs[0]], inst_props.error)
```

Should it be this instead?

File: releasenotes/notes/update-plot-gate-map-9ed6ad5490bafbbf.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+---
+features:
+  - |
+    The :func:`~.plot_gate_map` visualization function and the functions built
+    on top of it, :func:`~.plot_error_map` and :func:`~.plot_circuit_layout`,
+    have a new keyword argument, ``qubit_coordinates``. This argument takes
+    a sequence of 2D coordinates to use for plotting each qubit in the backend
+    being visualized. If specified this sequence must have a length equal to
```
Comment (kevinhartman) [inline releasenotes/notes/update-plot-gate-map-9ed6ad5490bafbbf.yaml:8]: The doc strings don't seem to mention the new parameter being optional. Is it worth documenting in corresponding methods?

File: qiskit/visualization/gate_map.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/visualization/gate_map.py:515]: Oh, yep that's a good catch. Not sure why I put 1 here, it definitely is supposed to be 0 as this is trying to find negative values and shift everything up so it's all >=0.

File: qiskit/visualization/gate_map.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/visualization/gate_map.py:967]: Yep good call, as you can tell there isn't great test coverage on this code. It's something we should fix I think there is an issue related to this: https://github.com/Qiskit/qiskit-terra/issues/7691 although we should expand the coverage as part of this.

File: qiskit/visualization/gate_map.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/visualization/gate_map.py:945]: yeah there is no reason to have a continue here, I just copied this from what was here before, but I'll just drop this in the next revision

File: releasenotes/notes/update-plot-gate-map-9ed6ad5490bafbbf.yaml (hunk 4 above)
Comment (mtreinish) [reply releasenotes/notes/update-plot-gate-map-9ed6ad5490bafbbf.yaml:8]: I kind of just assumed people would see it was a kwarg with a default of None they'd know it was optional. But it's good to be explicit in the docstring, I updated it here: https://github.com/Qiskit/qiskit-terra/pull/7814/commits/5b2d5ecb2ee759dcb7b99d4df83f1e5333bb80ec

File: qiskit/visualization/gate_map.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/visualization/gate_map.py]: Dropped in: https://github.com/Qiskit/qiskit-terra/pull/7814/commits/fac6ffb169a60e147e0340c68c2f570fdcffcc1b

File: qiskit/visualization/gate_map.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/visualization/gate_map.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/7814/commits/01fb3a03e30f59338c032f085ceb8314ed7a4ea8

**Review by kevinhartman — APPROVED** [review]

**Pull request merged**